### ISSUE TITLE
Return the highest complexity of nested fragments when using a UnionType or Interfaces

### ISSUE THREAD

**Issue by dmateiu** [slicknode/graphql-query-complexity#26]
When using UnionType or Interfaces, the current implementation is returning the sum of all nested fragments. 

As the types of the conditional fragments are mutually exclusive, I would not expect the complexities to be added up.

I would suggest using the complexity of the highest nested fragment.

Example:

```ts
      query {
        union {
          ...on Item { // assume complexity 22
            scalar
            complexScalar
          }
          ...on SecondItem { // assume complexity 2
            name
            scalar
          }
         value // assume complexity 1
        }
      }
    `
```

I would expect the resulting complexity to be 23.

**Comment by ivome** [comment]
This would be a great addition to add support for interfaces / unions. 

> I would suggest using the complexity of the highest nested fragment.

I think if we add support for this, we should add it in a way that covers all use cases. We cannot just return the complexity of the highest nested fragment because you can also add fragments of interfaces and a single type can implement multiple interfaces. 

For example:

Schema
```graphql
union Animal = Cat | Fish | Bird

interface AnimalInterface {
  name: String
}

interface Mammal {
  legs: Int
}

interface LandAnimal {
  canSwim: Boolean
}

type Cat implements Mammal & LandAnimal & AnimalInterface {
  legs: Int
  canSwim: Boolean
  name: String
}

type Fish & AnimalInterface {
  name: String
}

type Bird implements LandAnimal & AnimalInterface {
  legs: Int
  canSwim: Boolean
  name: String
}

type Query {
  animals: [Animal!]!
}
```

Query:

You can then write a query like this
```graphql
{
  animals {
    ...on LandAnimal {
      canSwim
    }
    ...on Mammal {
       legs
    }
  }
}
```

For a `Cat` you would have to add the complexities of both interfaces. We should add them up for each possible object type and then return the highest possible value as the resulting complexity. 

I'd happily accept a PR if you want to give it a shot...

**Comment by dmateiu** [comment]
> I think if we add support for this, we should add it in a way that covers all use cases. We cannot just return the complexity of the highest nested fragment because you can also add fragments of interfaces and a single type can implement multiple interfaces.

You are right.

I will dig into the code and reach out to you. I'd be happy to contribute.

**Comment by ivome** [comment]
> I will dig into the code and reach out to you. I'd be happy to contribute.

Great! In graphql-js 15.0.0 there is also a new feature that allows interfaces to extend other interfaces. This might be relevant for the implementation: https://github.com/graphql/graphql-js/pull/2084

**Comment by ivome** [comment]
This is published in version v0.7.0
